File: src/js/classes/petAura.js
class PetAura {
  constructor (pet) {
    this.pet = pet
    this.duration = 0
    this.durationRemaining = 0
    this.active = false
  }

  tick (t) {
    if (this.active) {
      this.durationRemaining = Math.max(0, this.durationRemaining - t)
      if (this.durationRemaining == 0) {
        this.fade()
      }
    }
  }

  apply () {
    this.active = true
    this.durationRemaining = this.duration
    let combatLogMsg = this.name + ' gains ' + this.name
    if (this.stacks) combatLogMsg += '(' + this.stacks + '). Current AP: ' + Math.round(this.pet.getAttackPower()) + ')'
    this.pet.player.combatLog(combatLogMsg)
  }

  fade (endOfIteration = false) {
    this.active = false
    this.durationRemaining = 0
    if (!endOfIteration) {
      let combatLogMsg = this.name
      if (this.stacks) combatLogMsg += '(' + this.stacks + ')'
      combatLogMsg += ' fades from ' + this.pet.name
      this.pet.player.combatLog(combatLogMsg)
    }
  }
}

class DemonicFrenzy extends PetAura {
  constructor (pet) {
    super(pet)
    this.name = 'Demonic Frenzy'
    this.duration = 10
    this.maxStacks = 10
    this.stacks = 0
  }

  apply () {
    if (this.stacks < this.maxStacks) {
      this.stacks++
    }
    super.apply()
  }

  fade (endOfIteration = false) {
    this.stacks = 0
    super.fade(endOfIteration)
  }
}

class BlackBook extends PetAura {
  constructor (pet) {
    super(pet)
    this.name = 'Black Book'
    this.duration = 30
  }

  apply(announceInCombatlog = true) {
    if (!this.active) {
      super.apply()
      this.pet.player.combatLog(this.pet.name + ' Spell Power + 200 (' + Math.round(this.pet.stats.spellPower) + ' -> ' + Math.round(this.pet.stats.spellPower + 200) + ')')
      this.pet.player.combatLog(this.pet.name + ' Attack Power + 325 (' + Math.round(this.pet.stats.ap) + ' -> ' + Math.round(this.pet.stats.ap + 325) + ')')
      this.pet.stats.buffs.spellPower += 200
      this.pet.stats.buffs.ap += 325
      this.pet.calculateStatsFromPlayer(announceInCombatlog)
    }
  }

  fade(endOfIteration = false) {
    if (this.active) {
      super.fade(endOfIteration)
      this.pet.player.combatLog(this.pet.name + ' Spell Power - 200 (' + Math.round(this.pet.stats.spellPower) + ' -> ' + Math.round(this.pet.stats.spellPower - 200) + ')')
      this.pet.player.combatLog(this.pet.name + ' Attack Power - 325 (' + Math.round(this.pet.stats.ap) + ' -> ' + Math.round(this.pet.stats.ap - 325) + ')')
      this.pet.stats.buffs.spellPower -= 200
      this.pet.stats.buffs.ap -= 325
      this.pet.calculateStatsFromPlayer(!endOfIteration)
    }
  }
}

File: src/cpp/pet_aura.cc
#include "pet_aura.h"

#include "common.h"
#include "player.h"

PetAura::PetAura(std::shared_ptr<Pet> pet)
    : pet(pet), duration(0), duration_remaining(0), active(false), stacks(0), max_stacks(0) {}

void PetAura::Tick(double t) {
  if (!active) {
    pet->player.ThrowError("Error: Ticking " + name + " when it isn't active");
  }

  duration_remaining -= t;

  if (duration_remaining <= 0) {
    Fade();
  }
}

void PetAura::Apply() {
  duration_remaining = duration;

  if (pet->player.ShouldWriteToCombatLog()) {
    std::string msg = pet->name + " " + (active && stacks == max_stacks ? "refreshed" : "gains") + " " + name;

    if (pet->auras.demonic_frenzy != NULL && name == pet->auras.demonic_frenzy->name) {
      msg.append(" (" + std::to_string(stacks) + ") - Current AP: " + DoubleToString(round(pet->GetAttackPower())) +
                 ")");
    }

    pet->player.CombatLog(msg);
  }

  for (auto& stat : stats) {
    stat.AddStat();
  }

  if (stacks < max_stacks) {
    stacks++;
  }
  active = true;
  pet->CalculateStatsFromPlayer();
}

void PetAura::Fade() {
  if (!active) {
    pet->player.ThrowError("Error: Trying to fade " + name + " when it isn't active");
  }

  active = false;
  duration_remaining = 0;
  stacks = 0;

  if (pet->player.ShouldWriteToCombatLog()) {
    pet->player.CombatLog(name + " faded");
  }

  for (auto& stat : stats) {
    stat.RemoveStat();
  }

  pet->CalculateStatsFromPlayer();
}

DemonicFrenzy::DemonicFrenzy(std::shared_ptr<Pet> pet) : PetAura(pet) {
  name = "Demonic Frenzy";
  duration = 10;
  max_stacks = 10;
}

BlackBook::BlackBook(std::shared_ptr<Pet> pet) : PetAura(pet) {
  name = "Black Book";
  duration = 30;
  stats = std::vector<Stat>{SpellPower(pet->player, pet->buff_stats, EntityType::kPet, 200),
                            AttackPower(pet->player, pet->buff_stats, EntityType::kPet, 325)};
}

BattleSquawk::BattleSquawk(std::shared_ptr<Pet> pet) : PetAura(pet) {
  name = "Battle Squawk";
  duration = 300;
  stats = std::vector<Stat>{MeleeHastePercent(pet->player, pet->stats, EntityType::kPet,
                                              std::pow(1.05, pet->player.settings.battle_squawk_amount))};
}